File: backend/src/services/parser.ts
import fs from 'fs';
import path from 'path';
import { md5 } from '../utils/crypto';
import { UserInfo } from '../types';

export class WeChatParser {
  /**
   * 解析 LoginInfo2.dat 文件，提取微信号和昵称
   * 核心算法从旧代码 parseLoginInfo 迁移
   */
  parseLoginInfo(documentsPath: string): Map<string, UserInfo> {
    const userInfoMap = new Map<string, UserInfo>();
    const loginInfoPath = path.join(documentsPath, 'LoginInfo2.dat');

    if (!fs.existsSync(loginInfoPath)) {
      console.log('LoginInfo2.dat 不存在');
      return userInfoMap;
    }

    try {
      const buffer = fs.readFileSync(loginInfoPath);
      const content = buffer.toString('binary');

      // 查找所有 wxid_ 模式的微信号
      const wxidPattern = /(wxid_[a-z0-9]{10,20})/g;
      const matches = content.match(wxidPattern);

      if (!matches) {
        console.log('未找到任何微信号');
        return userInfoMap;
      }

      // 去重
      const uniqueWxids = Array.from(new Set(matches));
      console.log(`LoginInfo2.dat 中找到 ${uniqueWxids.length} 个微信号:`, uniqueWxids);

      // 为每个微信号提取昵称
      for (const wxid of uniqueWxids) {
        const wxidIndex = content.indexOf(wxid);

        if (wxidIndex > -1) {
          // 在微信号后面的 100 个字节内查找昵称
          const afterWxid = content.substring(wxidIndex + wxid.length, wxidIndex + wxid.length + 100);
          const possibleNames: string[] = [];
          let currentName = '';

          // 提取可打印字符串
          for (let k = 0; k < afterWxid.length; k++) {
            const charCode = afterWxid.charCodeAt(k);
            // ASCII 可打印字符或中文字符
            if ((charCode >= 0x20 && charCode <= 0x7E) || charCode >= 0x4E00) {
              currentName += afterWxid[k];
            } else {
              if (currentName.length >= 2 && currentName.length <= 30) {
                possibleNames.push(currentName);
              }
              currentName = '';
            }
          }

          // 过滤出最可能的昵称
          let nickname = '';
          for (const name of possibleNames) {
            const trimmed = name.trim();
            // 排除全大写、纯数字、纯符号的字符串
            if (
              trimmed &&
              !/^[A-Z_]+$/.test(trimmed) &&
              !/^\+?\d+$/.test(trimmed) &&
              !/^[^a-zA-Z0-9\u4e00-\u9fa5]+$/.test(trimmed)
            ) {
              nickname = trimmed;
              break;
            }
          }

          // 计算 MD5
          const wxidMd5 = md5(wxid);
          userInfoMap.set(wxidMd5, {
            md5: wxidMd5,
            wechatId: wxid,
            nickname: nickname || wxid,
          });
        }
      }

      console.log(`成功解析 ${userInfoMap.size} 个用户信息`);
      return userInfoMap;
    } catch (error) {
      console.error('parseLoginInfo 错误:', error);
      return userInfoMap;
    }
  }

  /**
   * 扫描 documents 目录，查找所有用户目录
   */
  scanUsers(documentsPath: string): string[] {
    const userMd5s: string[] = [];

    if (!fs.existsSync(documentsPath)) {
      console.log('微信数据目录不存在:', documentsPath);
      return userMd5s;
    }

    const items = fs.readdirSync(documentsPath);

    for (const item of items) {
      const itemPath = path.join(documentsPath, item);
      const stat = fs.statSync(itemPath);

      // 检查是否是 32 位 MD5 格式的目录
      if (stat.isDirectory() && /^[a-f0-9]{32}$/i.test(item)) {
        // 检查是否包含 DB 目录
        const dbPath = path.join(itemPath, 'DB');
        if (fs.existsSync(dbPath) && fs.statSync(dbPath).isDirectory()) {
          userMd5s.push(item.toLowerCase());
        }
      }
    }

    console.log(`找到 ${userMd5s.length} 个用户目录`);
    return userMd5s;
  }

  /**
   * 获取用户头像路径
   */
  getUserAvatar(documentsPath: string, userMd5: string): string | undefined {
    const avatarPath = path.join(documentsPath, userMd5, 'Avatar', 'lastHeadImage');
    return fs.existsSync(avatarPath) ? avatarPath : undefined;
  }
}

File: backend/app/services/parser.py
"""微信数据解析服务"""
import os
import re
from pathlib import Path
from typing import Dict, List, Optional

from app.models import UserInfo
from app.utils.crypto import md5


class WeChatParser:
    """微信数据解析器"""
    
    def parse_login_info(self, documents_path: str) -> Dict[str, UserInfo]:
        """
        解析 LoginInfo2.dat 文件，提取微信号和昵称
        
        Args:
            documents_path: 微信数据目录路径
            
        Returns:
            用户信息字典 {md5: UserInfo}
        """
        user_info_map: Dict[str, UserInfo] = {}
        login_info_path = Path(documents_path) / 'LoginInfo2.dat'
        
        if not login_info_path.exists():
            print('LoginInfo2.dat 不存在')
            return user_info_map
        
        try:
            # 以二进制模式读取
            with open(login_info_path, 'rb') as f:
                content = f.read()
            
            # 转换为 latin-1 编码的字符串（保留所有字节）
            content_str = content.decode('latin-1')
            
            # 查找所有 wxid_ 模式的微信号
            wxid_pattern = re.compile(r'(wxid_[a-z0-9]{10,20})')
            matches = wxid_pattern.findall(content_str)
            
            if not matches:
                print('未找到任何微信号')
                return user_info_map
            
            # 去重
            unique_wxids = list(set(matches))
            print(f'LoginInfo2.dat 中找到 {len(unique_wxids)} 个微信号: {unique_wxids}')
            
            # 为每个微信号提取昵称
            for wxid in unique_wxids:
                wxid_index = content_str.find(wxid)
                
                if wxid_index > -1:
                    # 在微信号后面的 100 个字节内查找昵称
                    after_wxid = content_str[wxid_index + len(wxid):wxid_index + len(wxid) + 100]
                    possible_names: List[str] = []
                    current_name = ''
                    
                    # 提取可打印字符串
                    for char in after_wxid:
                        char_code = ord(char)
                        # ASCII 可打印字符或中文字符
                        if (0x20 <= char_code <= 0x7E) or char_code >= 0x4E00:
                            current_name += char
                        else:
                            if 2 <= len(current_name) <= 30:
                                possible_names.append(current_name)
                            current_name = ''
                    
                    # 过滤出最可能的昵称
                    nickname = ''
                    for name in possible_names:
                        trimmed = name.strip()
                        # 排除全大写、纯数字、纯符号的字符串
                        if (trimmed and 
                            not re.match(r'^[A-Z_]+$', trimmed) and
                            not re.match(r'^\+?\d+$', trimmed) and
                            not re.match(r'^[^a-zA-Z0-9\u4e00-\u9fa5]+$', trimmed)):
                            nickname = trimmed
                            break
                    
                    # 计算 MD5
                    wxid_md5 = md5(wxid)
                    user_info_map[wxid_md5] = UserInfo(
                        md5=wxid_md5,
                        wechatId=wxid,
                        nickname=nickname or wxid
                    )
            
            print(f'成功解析 {len(user_info_map)} 个用户信息')
            return user_info_map
        except Exception as e:
            print(f'parseLoginInfo 错误: {e}')
            return user_info_map
    
    def scan_users(self, documents_path: str) -> List[str]:
        """
        扫描 documents 目录，查找所有用户目录
        
        Args:
            documents_path: 微信数据目录路径
            
        Returns:
            用户 MD5 列表
        """
        user_md5s: List[str] = []
        documents_dir = Path(documents_path)
        
        if not documents_dir.exists():
            print(f'微信数据目录不存在: {documents_path}')
            return user_md5s
        
        # 遍历目录
        for item in documents_dir.iterdir():
            if item.is_dir():
                # 检查是否是 32 位 MD5 格式的目录
                if re.match(r'^[a-f0-9]{32}$', item.name, re.IGNORECASE):
                    # 检查是否包含 DB 目录
                    db_path = item / 'DB'
                    if db_path.exists() and db_path.is_dir():
                        user_md5s.append(item.name.lower())
        
        print(f'找到 {len(user_md5s)} 个用户目录')
        return user_md5s
    
    def get_user_avatar(self, documents_path: str, user_md5: str) -> Optional[str]:
        """
        获取用户头像路径
        
        Args:
            documents_path: 微信数据目录路径
            user_md5: 用户 MD5
            
        Returns:
            头像路径，不存在则返回 None
        """
        avatar_path = Path(documents_path) / user_md5 / 'Avatar' / 'lastHeadImage'
        return str(avatar_path) if avatar_path.exists() else None

